main/lsp/LSPTypechecker.h

```diff
@@ -27,23 +31,44 @@ class TypecheckRun final {
     LSPFileUpdates updates;
     // Specifies if the typecheck run took the fast or slow path.
     bool tookFastPath = false;
-    // Specifies if the typecheck run was canceled.
-    bool canceled = false;
     // If update took the slow path, contains a new global state that should be used moving forward.
     std::optional<std::unique_ptr<core::GlobalState>> newGS;
 
     TypecheckRun(std::vector<std::unique_ptr<core::Error>> errors = {},
                  std::vector<core::FileRef> filesTypechecked = {}, LSPFileUpdates updates = {},
                  bool tookFastPath = false, std::optional<std::unique_ptr<core::GlobalState>> newGS = std::nullopt);
-
-    // Make a canceled TypecheckRun.
-    static TypecheckRun makeCanceled();
 };
 
 /**
  * Encapsulates typechecker operations and enforces that they happen on a single thread.
  */
 class LSPTypechecker final {
+    /**
+     * Contains the LSPTypechecker state that is needed to cancel a running slow path operation and any subsequent fast
+     * path operations that have preempted it.
+     */
+    class UndoState final {
+    public:
```

Comment:
can any of these be const?
can any of these be private?

Reply:
`const` -- no, because we `move` these items during restoration.

`private` -- yes, but only if I move `restore` logic to `UndoState`. That may actually be a good idea...